File: euler_084.py
from time import time
from random import randint, shuffle
from collections import deque, Counter

def roll():
    die1 = randint(1, 4)
    die2 = randint(1, 4)
    return (die1 + die2, die1 == die2)

def chance():
    text = ["Nothing"] * 6 + ["Go", "Jail", "C1", "E3", "H2", "R1", "Next R", "Next R", "Next U", "Back 3"]
    shuffle(text)
    return deque(text)

def community_chest():
    text = ["Nothing"] * 14 + ["Go", "Jail"]
    shuffle(text)
    return deque(text)

def process_text(text, pos):
    if text == "Nothing":
        return pos
    if text == "Go":
        return 0
    if text == "Jail":
        return 10
    if text == "C1":
        return 11
    if text == "E3":
        return 24
    if text == "H2":
        return 39
    if text == "R1":
        return 5
    if text == "Next R":
        if pos in [2, 36]:
            return 5
        if pos == 7:
            return 15
        if pos in [17, 22]:
            return 25
        if pos == 33:
            return 35
    if text == "Next U":
        if pos in [17, 22]:
            return 28
        else:
            return 12
    if text == "Back 3":
        return (pos + 40  - 3)%40

start = time()

#Setup decks
cc1 = community_chest()
cc2 = community_chest()
cc3 = community_chest()
ch1 = chance()
ch2 = chance()
ch3 = chance()

#Starting the simulation
pos = 0
visited = Counter()
last_two_rolls = []

#The simulation
for it in range(100000):

    new_roll = roll()

    new_pos = (pos + new_roll[0])%40

    if new_pos == 30:
        new_pos = 10

    if new_pos == 2:
        text = cc1.popleft()
        cc1.append(text)
        new_pos = process_text(text, new_pos)

    if new_pos == 17:
        text = cc2.popleft()
        cc2.append(text)
        new_pos = process_text(text, new_pos)

    if new_pos == 33:
        text = cc3.popleft()
        cc3.append(text)
        new_pos = process_text(text, new_pos)

    if new_pos == 7:
        text = ch1.popleft()
        ch1.append(text)
        new_pos = process_text(text, new_pos)

    if new_pos == 22:
        text = ch2.popleft()
        ch2.append(text)
        new_pos = process_text(text, new_pos)

    if new_pos == 36:
        text = ch3.popleft()
        ch3.append(text)
        new_pos = process_text(text, new_pos)

    #Checking for double rolls
    last_two_rolls = [new_roll[1]] + last_two_rolls
    if len(last_two_rolls) == 3 and all(last_two_rolls):
        new_pos = 10

    visited.update([new_pos])

    pos = new_pos

print(visited.most_common(5))

stop = time()
print("Time: " + str(stop-start))

File: euler_084.js
const start = Date.now();

// Initializing the board
const BOARD = ["GO","A1","CC1","A2","T1","R1","B1",
"CH1","B2","B3","JAIL","C1","U1","C2","C3","R2","D1",
"CC2","D2","D3","FP","E1","CH2","E2","E3","R3","F1",
"F2","U2","F3","G2J","G1","G2","CC3","G3","R4","CH3",
"H1","T2","H2"];

// Initializing the Community Chest 2/16
const CC = ["GO","JAIL","NOTHING","NOTHING","NOTHING"
,"NOTHING","NOTHING","NOTHING","NOTHING","NOTHING"
,"NOTHING","NOTHING","NOTHING","NOTHING","NOTHING","NOTHING"];

// Initializing the Chance Deck 10/16
const CHANCE = ["GO","JAIL","C1","E3","H2","R1","NEXTR",
"NEXTR","NEXTU","BACK3","NOTHING","NOTHING","NOTHING","NOTHING"
,"NOTHING","NOTHING"];

// Shuffler: Modern Fisher-Yates
const shuffle = (arr) => {
    const swap = (arr2, i1, i2) => {
        const e1 = arr2[i1];
        const e2 = arr2[i2];
        arr2[i1] = e2;
        arr2[i2] = e1;
        return arr2;
    }

    for (let i = arr.length - 1; i > 0; i--) {
        const j = Math.floor(Math.random() * (i + 1));
        arr = swap(arr, i, j);
    }

    return arr;
}

// Making a die
class Die {
  constructor(sides) {
    this.sides = sides;
  }

  roll() {
    return Math.floor(Math.random() * this.sides + 1);
  }
}

// Making a game
class Game {
    constructor(board, communityChest, chanceDeck, numberOfRounds) {
        this.board = board;
        this.cc = [shuffle(communityChest), shuffle(communityChest), shuffle(communityChest)];
        this.ch = [shuffle(chanceDeck),shuffle(chanceDeck),shuffle(chanceDeck)];
        this.rounds = numberOfRounds;
        this.dieOne = new Die(4);
        this.dieTwo = new Die(4);
        this.played = 0;
        this.ccPos = [0, 0, 0]; // Top card
        this.chPos = [0, 0, 0]; // Top card
        this.counter = new Array(40).fill(0);
    }

    getCounter() {
      return this.counter;
    }

    getPlayed() {
      return this.played;
    }

    // Returns the new board position
    processCommunityChest(card, current) {
      if (card === "GO" || card === "JAIL") {
        return this.board.indexOf(card);
      }
      return current;
    }

    processChanceDeck(card, current) {
      if (card === "GO" || card === "JAIL" || card === "C1" || card === "E3" || card === "H2" || card === "R1") {
        return this.board.indexOf(card);
      }
      if (card === "BACK3") {
        return (current - 3) % 40;
      }
      if (card === "NEXTR") {
        const railways = ["R1", "R2", "R3", "R4"].map(x => this.board.indexOf(x));
        if (current >= railways[0] && current < railways[1]) {
          return this.board.indexOf("R2");
        }
        if (current >= railways[1] && current < railways[2]) {
          return this.board.indexOf("R3");
        }
        if (current >= railways[2] && current < railways[3]) {
          return this.board.indexOf("R4");
        }
        return this.board.indexOf("R1");
      }
      if (card === "NEXTU") {
        const utilities = ["U1", "U2"].map(x => this.board.indexOf(x));
        if (current >= utilities[0] && current < utilities[1]) {
          return this.board.indexOf("U2");
        }
        return this.board.indexOf("U1");
      }
      return current;
    }

    runGame() {
      let current = 0; // Begin at GO
      let consec = 0;
      while (this.played < this.rounds) {
        let roundDone = false;
        this.played++;
        const roll1 = this.dieOne.roll();
        const roll2 = this.dieTwo.roll();
        if (roll1 === roll2) {
          consec++;
        } else {
          consec = 0;
        }
        // Check for three consecutive rolls
        if (consec === 3) {
          current = this.board.indexOf("JAIL");
          consec = 0;
          roundDone = true;
        }

        if (!roundDone) {
          current = (current + roll1 + roll2) % 40;
        }
        if (!roundDone && this.board[current] === "G2J") {
            current = this.board.indexOf("JAIL");
            roundDone = true;
            this.counter[current]++;
        }
        if (!roundDone && this.board[current] === "CH1") {
          current = this.processChanceDeck(this.ch[0][this.chPos[0]], current);
          this.chPos[0] = (this.chPos[0] + 1) % 16;
        }
        if (!roundDone && this.board[current] === "CH2") {
          current = this.processChanceDeck(this.ch[1][this.chPos[1]], current);
          this.chPos[1] = (this.chPos[1] + 1) % 16;
        }
        if (!roundDone && this.board[current] === "CH3") {
          current = this.processChanceDeck(this.ch[2][this.chPos[2]], current);
          this.chPos[2] = (this.chPos[2] + 1) % 16;
        }
        if (!roundDone && this.board[current] === "CC1") {
          current = this.processCommunityChest(this.cc[0][this.ccPos[0]], current);
          this.ccPos[0] = (this.ccPos[0] + 1) % 16;
        }
        if (!roundDone && this.board[current] === "CC2") {
          current = this.processCommunityChest(this.cc[1][this.ccPos[1]], current);
          this.ccPos[1] = (this.ccPos[1] + 1) % 16;
        }
        if (!roundDone && this.board[current] === "CC3") {
          current = this.processCommunityChest(this.cc[2][this.ccPos[2]], current);
          this.ccPos[2] = (this.ccPos[2] + 1) % 16;
        }
        if (!roundDone) {
          this.counter[current]++;
        }
      }
    }
  }

const game = new Game(BOARD, shuffle(CC), shuffle(CHANCE), 1000000);

game.runGame();
const counts = game.getCounter();

let scounts = [...counts];
scounts.sort((a,b) => a - b);
let popular1 = counts.indexOf(scounts[39]);
let popular2 = counts.indexOf(scounts[38]);
let popular3 = counts.indexOf(scounts[37]);

const answer = popular1.toString().concat(popular2.toString()).concat(popular3.toString());


const end = Date.now();

console.log("Execute time: ", end - start, " The modal string is: ", answer);